### PR TITLE
#163050194 Implement JWT Authentication

### PR DESCRIPTION
**Pull request by VivianDoreen**
**What does this PR do?**
This PR delivers a bug that  is to ensure authentication is properly configured

**Description of Task to be completed?**
Implement the JWT Authentication

**How should this be manually tested?**
After the login route is implemented, it should return JWT token.

**What are the relevant pivotal tracker stories?**
[#163050194](https://www.pivotaltracker.com/story/show/163050194)

### CODE REVIEW

**Review by BarnaTB — APPROVED** [review]
LGTM. Great work Vivian.

**Comment by archbishop12** [comment]
@VivianDoreen Please provide some screenshot in your pr description and fix the commit message.

**Review by myrdstom — APPROVED** [review]
My opinion is this can only be fully tested during implementation of the  story [#162685658](https://www.pivotaltracker.com/n/projects/2232496). We should keep track of this bug

File: authors/apps/authentication/backends.py [hunk 1]
```diff
@@ -1,10 +1,46 @@
...
 """Configure JWT Here"""
+class JWTAuthentication(authentication.BaseAuthentication):
+    authentication_header_prefix = 'token'
+    def authenticate(self, request):
+        request.user = None
+        
+        auth_header = authentication.get_authorization_header(request).split()
+        auth_header_prefix = self.authentication_header_prefix.lower()
+
+        if not auth_header:
```
Comment (archbishop12) [inline authors/apps/authentication/backends.py]: @VivianDoreen You can refactor this code and reduce the number of if's that are used. Look into this.

File: authors/apps/authentication/backends.py [hunk 2]
```diff
@@ -1,10 +1,46 @@
...
+            return None
+        return self._authenticate_credentiatials(request, token)
+    
+    def _authenticate_credentiatials(self, request, token):
+        try:
+            payload = jwt.decode(token, settings.SECRET_KEY)
+        except:
+            msg = 'Invalid authentication. Could not decode token'
+            raise exceptions.AuthenticationFailed(msg)
+        try:
```
Comment (archbishop12) [inline authors/apps/authentication/backends.py]: @VivianDoreen Consider having a look at all the try blocks here and if they are necessary and refactor them  or you leave a comment stating why you needed all the stated scenarios.

File: authors/apps/authentication/backends.py [hunk 3]
```diff
@@ -1,10 +1,43 @@
...
+            prefix = auth_header[0].decode('utf-8')
+            token = auth_header[1].decode('utf-8')
+            if prefix.lower() != auth_header_prefix:
+                msg = 'Bearer is expected as the prefix '
+                raise exceptions.AuthenticationFailed(msg)
+            else:
+                return self._authenticate_credentiatials(request, token)
+        else:
+            msg = 'The header must contain only two elements'
+            raise exceptions.AuthenticationFailed(msg)
```
Comment (archbishop12) [inline authors/apps/authentication/backends.py]: @VivianDoreen Change this to fix the problem of authenticating open routes.
```suggestion
            return None
```

File: authors/apps/authentication/backends.py [hunk 4]
```diff
@@ -1,10 +1,43 @@
...
+        if len(auth_header) == 2:
+            prefix = auth_header[0].decode('utf-8')
+            token = auth_header[1].decode('utf-8')
+            if prefix.lower() != auth_header_prefix:
+                msg = 'Bearer is expected as the prefix '
+                raise exceptions.AuthenticationFailed(msg)
+            else:
+                return self._authenticate_credentiatials(request, token)
+        else:
+            msg = 'The header must contain only two elements'
```
Comment (archbishop12) [inline authors/apps/authentication/backends.py]: @VivianDoreen Remove this message

File: authors/apps/authentication/backends.py (hunk 2 above)
Comment (VivianDoreen) [reply authors/apps/authentication/backends.py]: ok, let me do that

File: authors/apps/authentication/backends.py (hunk 3 above)
Comment (VivianDoreen) [reply authors/apps/authentication/backends.py]: Ok, let me do that

Comment (VivianDoreen) [reply authors/apps/authentication/backends.py]: I have made the necessary changes. kindly review

**Review by archbishop12 — APPROVED** [review]
@VivianDoreen I am merging this despite the fact that coverage has reduced, this will be tested out when one works on an endpoint that is authenticated.

**Pull request merged**